src/lib/ppx_coda/lint_version_syntax.ml

```diff
@@ -0,0 +1,139 @@
+
+*)
+
+open Core_kernel
+open Ppxlib
+
+let name = "enforce_version_syntax"
+
+let errors_as_warnings_ref = ref false
+
+let is_ident ident item =
+  let version_id id =
+    match id.txt with Lident s -> String.equal s ident | _ -> false
+  in
+  match item with
+  | Pexp_ident id ->
+      version_id id
+  | Pexp_apply ({pexp_desc= Pexp_ident id; _}, _) ->
+      version_id id
+  | _ ->
+      false
+
+let payload_has_item is_item_ident payload =
+  match payload with
+  | PStr structure ->
+      List.exists structure ~f:(fun str ->
+          match str.pstr_desc with
+          | Pstr_eval (expr, _) -> (
+            (* the "ident" can appear as a singleton ident, or in a tuple *)
+            match expr.pexp_desc with
+            | Pexp_ident _ ->
+                is_item_ident expr.pexp_desc
+            | Pexp_apply ({pexp_desc; _}, _) ->
+                is_item_ident pexp_desc
+            | Pexp_tuple items ->
+                List.exists items ~f:(fun item -> is_item_ident item.pexp_desc)
+            | _ ->
+                false )
+          | _ ->
+              false )
+  | _ ->
+      false
+
+let is_version_ident = is_ident "version"
+
+let payload_has_version = payload_has_item is_version_ident
+
+let is_bin_io_ident = is_ident "bin_io"
+
+let payload_has_bin_io = payload_has_item is_bin_io_ident
+
+let attribute_has_deriving_version ((name, payload) : attribute) =
+  String.equal name.txt "deriving" && payload_has_version payload
+
+let attributes_have_deriving_version (attrs : attribute list) =
+  List.exists attrs ~f:attribute_has_deriving_version
+
+let type_has_deriving_version type_decl =
+  attributes_have_deriving_version type_decl.ptype_attributes
+
+let is_deriving name = String.equal name "deriving"
+
+let make_deriving_validator ~pred err_msg type_decl =
+  let get_deriving_error (name, payload) =
+    if is_deriving name.txt then
+      let has_bin_io = payload_has_bin_io payload in
+      let has_version = payload_has_version payload in
+      if pred has_bin_io has_version then Some (name.loc, err_msg) else None
+    else None
+  in
+  List.filter_map type_decl.ptype_attributes ~f:get_deriving_error
+
+let validate_neither_bin_io_nor_version =
+  make_deriving_validator
+    ~pred:(fun has_bin_io has_version -> has_bin_io || has_version)
+    "Deriving bin_io and deriving version disallowed for types in functor body"
+
+let validate_version_if_bin_io =
+  make_deriving_validator
+    ~pred:(fun has_bin_io has_version -> has_bin_io && not has_version)
+    "Must have deriving version if deriving bin_io"
+
+(* traverse AST, collect errors *)
+let check_deriving_usage =
+  object (self)
+    (* bool indicates whether we're in a functor *)
+    inherit [bool * (Location.t * string) list] Ast_traverse.fold as super
+
+    method! module_expr expr ((in_functor, errors) as acc) =
+      match expr.pmod_desc with
+      | Pmod_functor (_label, _mty, body) ->
```

Comment:
On reflection, I think this should be `Pmod_functor (_label, Some _mty, body)` instead. We can guarantee that a versioned type in a functor with a `None` argument for the module type (like the below) only depends on its parameters, since there are certainly no versioned types the can 'change out from underneath it' when the functor is applied:
```ocaml
module M () = struct
  module Stable = struct
     ...
  end
end
```